xrviz/dashboard.py

```diff
@@ -0,0 +1,93 @@
+import panel as pn
+import xarray as xr
+import hvplot.xarray
+from cartopy import crs as ccrs
+from .sigslot import SigSlot
+from .control import Control
+from .utils import convert_widget
+
+
+class Dashboard(SigSlot):
+    """
+    This section provides access to the complete generated dashboard,
+    consisting of all subsections.
+
+    Parameters
+    ----------
+    data: `xarray` instance: `DataSet` or `DataArray`
+           datset is used to initialize.
+
+    Attributes
+    ----------
+    panel: Displays the generated dashboard.
+    control: Provides access to the generated control panel.
+    index_selectors: Provides access to generated index_selectors.
+    plot: Plot button, upon click generates graph according to
+          kwargs selected in other sub-sections.
+    output: Provides access to generated graph.
+    """
+
+    def __init__(self, data):
+        super().__init__()
+        self.data = data
+        self.control = Control(self.data)
+        self.plot = pn.widgets.Button(name='Plot', width=200)
+        self.index_selectors = self.control.fields.index_selectors
+        self.output = pn.Row(pn.Spacer(name='Graph'))
+
+        self._register(self.plot, 'plot_clicked', 'clicks')
+        self.connect('plot_clicked', self.create_plot)
+        self.control.displayer.connect('variable_selected', self.check_is_plottable)
+
+        self.panel = pn.Column(self.control.panel,
+                               self.plot,
+                               self.output
+                               )
+        if isinstance(self.data, xr.Dataset):
+            self.plot.disabled = True
+        else:
+            if self.data.name in self.data.coords or len(self.data.dims) <= 1:
+                self.plot.disabled = True
+
+    def create_plot(self, *args):
+        kwargs = self.control.kwargs
+        var = kwargs['']
+        graph_opts = {}
+        graph_opts['x'] = kwargs['x']
+        graph_opts['y'] = kwargs['y']
+        graph_opts['rasterize'] = True
+        graph_opts['width'] = 600
+        graph_opts['height'] = 400
+        graph_opts['crs'] = ccrs.PlateCarree()
+
+        if isinstance(self.data, xr.Dataset):
+            graph_opts['title'] = var
+            self.graph = pn.Row(self.data[var].hvplot.quadmesh(**graph_opts))
+        else:
+            graph_opts['title'] = self.data.name
+            self.graph = pn.Row(self.data.hvplot.quadmesh(**graph_opts))
+
+        self.output[0] = self.graph[0][0]
+        self.fill_index_selectors()
+
+    def fill_index_selectors(self):
+        """
+        To convert the auto-generated slider, into Select and Player
+        widget.
+        """
+        self.index_selectors.clear()
+        try:
+            for widget in self.graph[0][1]:
+                selector = convert_widget(widget, pn.widgets.Select())
```

Comment:
So the logic is to let holoviews take all the dimensions, create it's own sliders and then hijack their functionality. This is a fascinating way to go about it!

We can talk about this, but I think it would be better to make our own selection of the data before passing it to holoviews, and so cut out the magical hidden sliders.

Reply:
We will discuss more about this in the e-meet. Holoviews is definitely good in handling selection, so may be we should rely upon it, rather than creating same thing again. I might not offer much advantage besides cutting out sliders.

Reply:
True, but on the other hand, here we are relying in the detail of the holoviews implementation, and we don't know how reliable that is; also, we are causing each user action to trigger multiple events, which may affect the interface's responsiveness.

Reply:
This has been resolved now. We are creating our own selectors.